File: microservices/assortments/models/supplierProduct.py
from database.db import db
from flask_mongoengine import BaseQuerySet
from models.category import Category
from models.supplierCategory import SupplierCategory
import datetime
# from slugify import slugify, Slugify, UniqueSlugify


class SupplierProduct(db.Document):
    tenant_name = db.StringField(required=True)
    host_id = db.StringField(required=True)
    tenant_id = db.StringField(required=True)
    category_name = db.StringField(required=True)
    category_display_name = db.ListField(required=True, default=[])
    category_slug = db.StringField(required=True)
    linked = db.ReferenceField(Category)
    supplier_category = db.ReferenceField(SupplierCategory)
    shareable = db.BooleanField(default=True)
    published = db.BooleanField(default=True)
    object = db.ListField()
    runs = db.ListField(default=[])
    create_at = db.StringField(default='')
    created_at = db.DateTimeField(default=datetime.datetime.now, required=True)
    additional = db.DictField(required=False, default={})

    meta = {
        'indexes': ['-created_at','linked','supplier_category','tenant_id'],
        'ordering': ['-created_at'],
        'collection': 'supplier_products',
        'queryset_class': BaseQuerySet
    }

    def __setattr__(self, key, value):
        super(SupplierProduct, self).__setattr__(key, value)
        # if key == 'name':
        #     self.slug = slugify(self.name, to_lower=True)
    def __repr__(self):
        return '<SupplierProduct %r' % self.id

File: microservices/boxes/boxes/Models/SupplierProduct.js
const mongoose = require("mongoose");
const slugify = require("slugify");
const Schema = mongoose.Schema;
const ObjectId = Schema.Types.ObjectId

const SupplierProductSchema = Schema({
    tenant_name: { type:String, required: true},
    sort: {type: Number, required: false, default: 0},
    host_id: { type:String, required: true},
    tenant_id: {type:String, required: true, index: true},
    category_name: {type:String, required: true},
    category_display_name: { type:Array, required: true},
    category_slug: {type:String, required: true},
    linked: {type:ObjectId, required: false, index: true},
    supplier_category: {type:ObjectId, required: true, index: true},
    shareable: {type:Boolean, default: true},
    published: {type:Boolean, default: true},
    object: {type:Object, required: true},
    runs: {type:Array, required: true, default: []},
    additional: {type: Array, required: false, default: []},
    created_at: {type:Date, default: Date.now, index: true},
}, {
    versionKey: false // You should be aware of the outcome after set to false
});


// Pre-save hook to dynamically set the sort value and slug
SupplierProductSchema.pre('save', async function (next) {
    try {
        // Automatically set the slug if the name field is modified
        if (this.isModified('category_name')) {
            this.category_slug = slugify(this.category_name, { lower: true });
        }

        // Check if the document is new to dynamically set the sort value
        if (this.isNew) {
            const maxSortCategory = await this.constructor.findOne().sort({ sort: -1 });
            this.sort = maxSortCategory ? maxSortCategory.sort + 1 : 0;
        }

        next();
    } catch (err) {
        next(err);
    }
});

module.exports = SupplierProduct = mongoose.model("supplier_products", SupplierProductSchema);
